src/us/kbase/workspace/test/kbase/KBaseWorkspaceConfigTest.java

```diff
@@ -89,6 +121,295 @@ private void listenerConstructFail(
+		failLoadFromFile(notsofakefile, "thing", new KBaseWorkspaceConfigException(String.format(
+				"The configuration file %s/imnotfakeimnotfake has no section thing", TEMP_DIR)));
+		
+		final Path realfile = TEMP_DIR.resolve("keepinitreal");
+		Files.write(realfile, list("[Workspace]", "mongodb-host=foo"), StandardCharsets.UTF_8);
+		failLoadFromFile(realfile, null, new KBaseWorkspaceConfigException(String.format(
+				"The configuration file %s/keepinitreal has no section null", TEMP_DIR)));
+		failLoadFromFile(realfile, "Workspacey", new KBaseWorkspaceConfigException(String.format(
+				"The configuration file %s/keepinitreal has no section Workspacey", TEMP_DIR)));
+	}
+	
+	private void failLoadFromFile(
+			final Path file,
+			final String section,
+			final Exception expected) {
+		try {
+			new KBaseWorkspaceConfig(file, section);
+			fail("expected exception");
+		} catch (Exception got) {
+			TestCommon.assertExceptionCorrect(got, expected);
+		}
+	}
+	
+	private static class ExpectedConfig {
+		
+		public URL auth2URL = null;
+		public URL authURL = null;
+		public Set<String> adminReadOnlyRoles = set();
+		public Set<String> adminRoles = set();
+		public String workspaceAdmin = null;
+		public String mongohost = null;
+		public String mongoDBname = null;
+		public String typeDBname = null;
+		public String mongoPwd = null;
+		public String mongoUser = null;
+		public List<ListenerConfig> listenerConfigs = Collections.emptyList();
+		public BackendType backendType = null;
+		public URL backendURL = null;
+		public String backendUser = null;
+		public String backendToken = null;
+		public String backendContiner = null;
+		public Region backendRegion = null;
+		public boolean backendTrustAllCerts = false;
+		public URL handleServiceURL = null;
+		public String handleServiceToken = null;
+		public boolean ignoreHandleService = false;
+		public URL bytestreamURL = null;
+		public String bytestreamUser = null;
+		public String bytestreamToken = null;
+		public URL sampleServiceURL = null;
+		public String sampleServiceToken = null;
+		public String tempDir = null;
+		public List<String> infoMessages = Collections.emptyList();
+		public String paramReport = null;
+		public boolean hasErrors = false;
+		public List<String> errors = Collections.emptyList();
+
+		public ExpectedConfig withAdminReadOnlyRoles(final Set<String> adminReadOnlyRoles) {
+			this.adminReadOnlyRoles = adminReadOnlyRoles;
+			return this;
+		}
+
+		public ExpectedConfig withAdminRoles(final Set<String> adminRoles) {
+			this.adminRoles = adminRoles;
+			return this;
+		}
+
+		public ExpectedConfig withAuth2URL(final URL auth2url) {
+			auth2URL = auth2url;
+			return this;
+		}
+
+		public ExpectedConfig withAuthURL(final URL authURL) {
+			this.authURL = authURL;
+			return this;
+		}
+
+		public ExpectedConfig withBackendToken(final String backendToken) {
+			this.backendToken = backendToken;
+			return this;
+		}
```

Comment:
do all these declarations have to be written out in full? There's no shortcut where you write a function to generate a whole load of similar properties?

Reply:
Without using https://projectlombok.org/, which I haven't learned yet but is on my list, not without doing some really crazy stuff with the classloader AFAIK.